src/txrequest.cpp

```diff
@@ -0,0 +1,593 @@
+        // of the ByPeer index if a non-CANDIDATE_BEST announcement already exists with the same txhash and peer).
+        // Bail out in that case.
+        auto ret = m_index.get<ByPeer>().emplace(gtxid, peer, preferred, reqtime, m_current_sequence);
+        if (!ret.second) return;
+
+        // Update accounting metadata.
+        ++m_peerinfo[peer].m_total;
+        ++m_current_sequence;
+    }
+
+    //! Find the GenTxids to request now from peer.
+    std::vector<GenTxid> GetRequestable(NodeId peer, std::chrono::microseconds now)
+    {
+        // Move time.
+        SetTimePoint(now);
+
+        // Find all CANDIDATE_BEST announcements for this peer.
+        std::vector<const Announcement*> selected;
+        auto it_peer = m_index.get<ByPeer>().lower_bound(ByPeerView{peer, true, uint256::ZERO});
+        while (it_peer != m_index.get<ByPeer>().end() && it_peer->m_peer == peer &&
+            it_peer->m_state == State::CANDIDATE_BEST) {
+            selected.emplace_back(&*it_peer);
+            ++it_peer;
+        }
+
+        // Sort by sequence number.
+        std::sort(selected.begin(), selected.end(), [](const Announcement* a, const Announcement* b) {
+            return a->m_sequence < b->m_sequence;
+        });
+
+        // Convert to GenTxid and return.
+        std::vector<GenTxid> ret;
+        ret.reserve(selected.size());
+        std::transform(selected.begin(), selected.end(), std::back_inserter(ret), [](const Announcement* ann) {
+            return GenTxid{ann->m_is_wtxid, ann->m_txhash};
+        });
+        return ret;
+    }
+
+    void RequestedTx(NodeId peer, const uint256& txhash, std::chrono::microseconds expiry)
+    {
+        auto it = m_index.get<ByPeer>().find(ByPeerView{peer, true, txhash});
+        if (it == m_index.get<ByPeer>().end()) {
+            // There is no CANDIDATE_BEST announcement, look for a _READY or _DELAYED instead. If the caller only
+            // ever invokes RequestedTx with the values returned by GetRequestable, and no other non-const functions
+            // other than ForgetTxHash and GetRequestable in between, this branch will never execute (as txhashes
+            // returned by GetRequestable always correspond to CANDIDATE_BEST announcements).
+
+            it = m_index.get<ByPeer>().find(ByPeerView{peer, false, txhash});
+            if (it == m_index.get<ByPeer>().end() || (it->m_state != State::CANDIDATE_DELAYED &&
+                it->m_state != State::CANDIDATE_READY)) {
+                // There is no CANDIDATE announcement tracked for this peer, so we have nothing to do. Either this
+                // txhash wasn't tracked at all (and the caller should have called ReceivedInv), or it was already
+                // requested and/or completed for other reasons and this is just a superfluous RequestedTx call.
+                return;
+            }
+
+            // Look for an existing CANDIDATE_BEST or REQUESTED.
```

Comment:
Shouldn't this be outside the `if (it == m_index.get<ByPeer>().end()) {` condition and happen unconditionally?

Reply:
That's not necessary as it's only possible in case no CANDIDATE_BEST entry for the specified (peer, txhash) was found (if there was, then no other REQUESTED or CANDIDATE_BEST for the same txhash by another peer can exist, due to invariant that there is at most one of those per txhash). I've added more comments to hopefully clarify.

Reply:
Right.